pkg/fuzzer/fuzzer.go

```diff
@@ -68,67 +65,105 @@ func NewFuzzer(ctx context.Context, cfg *Config, rnd *rand.Rand,
-	Collide        bool
-	EnabledCalls   map[*prog.Syscall]bool
-	NoMutateCalls  map[int]bool
-	FetchRawCover  bool
-	NewInputFilter func(call string) bool
+type execQueues struct {
+	smashQueue           *queue.PlainQueue
+	triageQueue          *queue.PriorityQueue
+	candidateQueue       *queue.PlainQueue
+	triageCandidateQueue *queue.PriorityQueue
+	source               queue.Source
 }
 
-type Request struct {
-	Prog       *prog.Prog
-	NeedSignal SignalType
-	NeedCover  bool
-	NeedHints  bool
-	// If specified, the resulting signal for call SignalFilterCall
-	// will include subset of it even if it's not new.
-	SignalFilter     signal.Signal
-	SignalFilterCall int
-	// Fields that are only relevant within pkg/fuzzer.
-	flags   ProgTypes
-	stat    *stats.Val
-	resultC chan *Result
+func newExecQueues(fuzzer *Fuzzer) execQueues {
+	ret := execQueues{
+		triageCandidateQueue: queue.Priority(),
+		candidateQueue:       queue.PlainWithStat(fuzzer.StatCandidates),
+		triageQueue:          queue.Priority(),
+		smashQueue:           queue.Plain(),
+	}
+	// Sources are listed in the order, in which they will be polled.
+	ret.source = queue.Order(
+		ret.triageCandidateQueue,
+		ret.candidateQueue,
+		ret.triageQueue,
+		// Alternate smash jobs with exec/fuzz once in 3 times.
+		queue.Alternate(ret.smashQueue, 3),
+		queue.Callback(fuzzer.genFuzz),
+	)
+	return ret
 }
 
-type SignalType int
+type execOpt any
+type dontTriage struct{}
+type progFlags ProgTypes
 
-const (
-	NoSignal  SignalType = iota // we don't need any signal
-	NewSignal                   // we need the newly seen signal
-	AllSignal                   // we need all signal
-)
+func (fuzzer *Fuzzer) validateRequest(req *queue.Request) {
+	if req.NeedHints && (req.NeedCover || req.NeedSignal != queue.NoSignal) {
+		panic("Request.NeedHints is mutually exclusive with other fields")
+	}
+	if req.SignalFilter != nil && req.NeedSignal != queue.NewSignal {
+		panic("SignalFilter must be used with NewSignal")
+	}
+}
 
-type Result struct {
-	Info *ipc.ProgInfo
-	Stop bool
+func (fuzzer *Fuzzer) execute(executor queue.Executor, req *queue.Request, opts ...execOpt) *queue.Result {
+	fuzzer.validateRequest(req)
+	executor.Submit(req)
+	res := req.Wait(fuzzer.ctx)
+	fuzzer.processResult(req, res, opts...)
+	return res
 }
 
-func (fuzzer *Fuzzer) Done(req *Request, res *Result) {
+func (fuzzer *Fuzzer) prepare(req *queue.Request, opts ...execOpt) {
+	fuzzer.validateRequest(req)
+	req.OnDone(func(req *queue.Request, res *queue.Result) bool {
+		fuzzer.processResult(req, res, opts...)
+		return true
+	})
+}
+
+func (fuzzer *Fuzzer) enqueue(executor queue.Executor, req *queue.Request, opts ...execOpt) {
+	fuzzer.prepare(req, opts...)
+	executor.Submit(req)
+}
+
+func (fuzzer *Fuzzer) processResult(req *queue.Request, res *queue.Result, opts ...execOpt) {
+	var flags ProgTypes
+	var noTriage bool
+	for _, opt := range opts {
+		switch v := opt.(type) {
+		case progFlags:
+			flags = ProgTypes(v)
```

Comment:
We can use ProgTypes directly as an option, it's already a unique type.
And dontTriage looks like it belongs to ProgTypes as well. It already has similar dontMinimize and dontSmash flags. So it could be:
```
const (
	progCandidate ProgTypes = 1 << iota
	progMinimized
	progSmashed
	progTriaging
)
```
Then we don't need generic type-erased options.

Reply:
Done